### PR TITLE
Implement inference methods (predict, transform, and score)

### PR DESCRIPTION
**Pull request by fcharras**

### CODE REVIEW

**Comment by fcharras** [comment]
So this is the private API that was missing for inference methods (KMeans.transform. KMeans.predict, and KMeans.score) + unit tests.

Also I've dispatched the main kernels in several distinct files, it does not fix the issues with repeated code lines  but I think it feels lighter this way.

File: sklearn_numba_dpex/kmeans/kernels.py [hunk 1]
```diff
@@ -1,942 +0,0 @@
-import math
-from functools import lru_cache
-
-import dpctl
-import numba_dpex as dpex
-
-# General note on kernel implementation
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/kernels.py:7]: Is it possible to do another PR which would precedes this PR for the removal of this file and the migration of its content to the files in `sklearn_numba_dpex/kmeans/kernels`? This would reduce the diff of this PR and I think ease reviews (I think having at max ± 500 diffs is a good rule of thumb for faster integration).

**Review by jjerphan — COMMENTED** [review]
Thank you, @fcharras.

Here is a high level pass with general suggestions to make this PR easier to review.

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 2]
```diff
@@ -166,7 +170,7 @@ def __init__(
                     f"support it."
                 )
 
-    def __call__(
```
[the remark this answers is not in the document]
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/drivers.py:170]: Let's rename it KMeansDriver then ?

File: benchmark/kmeans.py [hunk 3]
```diff
@@ -217,14 +217,14 @@ def benchmark_data_initialization(random_state=random_state, n_clusters=n_cluste
...
             slow=True,
         )
 
         kmeans_timer.timeit(
             LLoydKMeansDriver(
                 device="gpu", dtype=dtype, work_group_size_multiplier=multiplier
-            ),
+            ).lloyd,
             name=f"Kmeans numba_dpex lloyd GPU (work_group_size_multiplier={multiplier})",
         )
```
[the remark this answers is not in the document]
Comment (fcharras) [reply benchmark/kmeans.py:230]: That is out of scope of this PR i'd prefer not address that here, it is expected that benchmarks will be iterated on later (adding new dataset, a better grid search, etc)

File: sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    def _initialize_window_of_centroids(
+        local_work_id,                  # PARAM
+        first_centroid_idx,             # PARAM
+        centroids_half_l2_norm,         # IN
+        centroids_window_half_l2_norm,  # OUT
+        dot_products,                   # OUT
+    ):
+    # fmt: on
+        # initialize the partial pseudo inertia for each of the window_n_centroids
+        # centroids in the window
```
[the remark this answers is not in the document]
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/_base_kmeans_kernel_funcs.py]: i'll apply the 3 following nits in #18 where it now belongs

File: sklearn_numba_dpex/kmeans/kernels.py (hunk 1 above)
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels.py]: While I agree with your rule of thumbs for diffs, i don't think it should apply here since the diff was artificially high because of copy/paste, even if they appear in the diff I don't think the internals of the copy/pasted blocks should be reviewed, they have not been edited, only moved left and right.

Even after rebasing on #18 it remains fairly high because of duplicated code that can't be refactored because known limitations with numba_dpex.

**Comment by fcharras** [comment]
Thank you for looking at this PR @jjerphan I've addressed your requests here and in #18

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 5]
```diff
@@ -416,6 +419,194 @@ def __call__(
...
+        X_t, centroids_t = self._load_transposed_data_to_device(X, centers)
+
+        centroids_half_l2_norm = dpctl.tensor.empty(
+            n_clusters, dtype=dtype, device=self.device
+        )
+        assignments_idx = dpctl.tensor.empty(
+            n_samples, dtype=np.uint32, device=self.device
+        )
+
+        half_l2_norm(centroids_t, centroids_half_l2_norm)
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:472]: This adds a `_kernel` suffix to `half_l2_norm` to match the kernels' naming convention.
```suggestion
        half_l2_norm_kernel = make_half_l2_norm_2d_axis0_kernel(
            size0=n_features,
            size1=n_clusters,
            work_group_size=work_group_size,
            dtype=dtype,
        )

        X_t, centroids_t = self._load_transposed_data_to_device(X, centers)

        centroids_half_l2_norm = dpctl.tensor.empty(
            n_clusters, dtype=dtype, device=self.device
        )
        assignments_idx = dpctl.tensor.empty(
            n_samples, dtype=np.uint32, device=self.device
        )

        half_l2_norm_kernel(centroids_t, centroids_half_l2_norm)
```

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 6]
```diff
@@ -416,6 +419,194 @@ def __call__(
...
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        reduce_inertia = make_sum_reduction_1d_kernel(
+            size=n_samples,
+            work_group_size=work_group_size,
+            device=self.device,
+            dtype=dtype,
+        )
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:529]: Same suggestion regarding adding a suffix for kernels' names.
```suggestion
        half_l2_norm_kernel = make_half_l2_norm_2d_axis0_kernel(
            size0=n_features,
            size1=n_clusters,
            work_group_size=work_group_size,
            dtype=dtype,
        )

        reduce_inertia_kernel = make_sum_reduction_1d_kernel(
            size=n_samples,
            work_group_size=work_group_size,
            device=self.device,
            dtype=dtype,
        )
```

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 7]
```diff
@@ -416,6 +419,194 @@ def __call__(
...
+        X_t, centroids_t = self._load_transposed_data_to_device(X, centers)
+
+        centroids_half_l2_norm = dpctl.tensor.empty(
+            n_clusters, dtype=dtype, device=self.device
+        )
+        per_sample_inertia = dpctl.tensor.empty(
+            n_samples, dtype=dtype, device=self.device
+        )
+
+        half_l2_norm(centroids_t, centroids_half_l2_norm)
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:540]:
```suggestion
        half_l2_norm_kernel(centroids_t, centroids_half_l2_norm)
```

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 8]
```diff
@@ -416,6 +419,194 @@ def __call__(
...
+
+        compute_inertia_fixed_window_kernel(
+            X_t,
+            centroids_t,
+            centroids_half_l2_norm,
+            per_sample_inertia,
+        )
+
+        inertia = dpctl.tensor.asnumpy(reduce_inertia(per_sample_inertia))[0]
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:549]: Why do we need to take the first returned element solely here?
What is the instruction returning?
```suggestion
        inertia = dpctl.tensor.asnumpy(reduce_inertia_kernel(per_sample_inertia))[0]
```

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 9]
```diff
@@ -416,6 +419,194 @@ def __call__(
...
+
+    def get_euclidean_distances(
+        self, X, Y=None, *, Y_norm_squared=None, squared=False, X_norm_squared=None
+    ):
+        """This call is expected to accept the same inputs than sklearn's private
+        euclidean_distances and returns euclidean distances of each sample to each
+        cluster center
+        """
+        if squared:
+            raise NotImplementedError("Only squared=False is allowed")
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:561]: Why is it the case exactly?

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 10]
```diff
@@ -416,6 +419,194 @@ def __call__(
...
+            n_features,
+            n_clusters,
+            preferred_work_group_size_multiple=self.preferred_work_group_size_multiple,
+            centroids_window_width_multiplier=self.centroids_window_width_multiplier,
+            centroids_window_height=self.centroids_window_height,
+            work_group_size=work_group_size,
+            dtype=dtype,
+        )
+
+        half_l2_norm = make_half_l2_norm_2d_axis0_kernel(
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:585]:
```suggestion
        half_l2_norm_kernel = make_half_l2_norm_2d_axis0_kernel(
```

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 11]
```diff
@@ -416,6 +419,194 @@ def __call__(
...
+        )
+
+        X_t, Y_t = self._load_transposed_data_to_device(X, Y)
+
+        Y_half_l2_norm = dpctl.tensor.empty(n_clusters, dtype=dtype, device=self.device)
+        euclidean_distances_t = dpctl.tensor.empty(
+            (n_clusters, n_samples), dtype=dtype, device=self.device
+        )
+
+        half_l2_norm(Y_t, Y_half_l2_norm)
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:599]:
```suggestion
        half_l2_norm_kernel(Y_t, Y_half_l2_norm)
```

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 12]
```diff
@@ -416,6 +419,194 @@ def __call__(
...
+        self,
+        X,
+        sample_weight,
+        x_squared_norms,
+        centers,
+        n_threads=1,
+        return_inertia=True,
+    ):
+        """This call is expected to accept the same inputs than sklearn's private
+        _labels_inertia and returns (labels, None)
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:432]:
```suggestion
        """This call is expected to accept the same inputs than
        `sklearn.cluster._kmeans._labels_inertia` while solely computing
        the samples' labels, hence returning (labels, None).
```

File: sklearn_numba_dpex/kmeans/drivers.py [hunk 13]
```diff
@@ -416,6 +419,194 @@ def __call__(
...
+        self,
+        X,
+        sample_weight,
+        x_squared_norms,
+        centers,
+        n_threads=1,
+        return_inertia=True,
+    ):
+        """This call is expected to accept the same inputs than sklearn's private
+        _labels_inertia and returns (None, inertia)
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/drivers.py:493]:
```suggestion
        """This call is expected to accept the same inputs than
        `sklearn.cluster._kmeans._labels_inertia` while solely computing
        the samples' inertia, hence returning (None, inertia).
```

**Review by jjerphan — COMMENTED** [review]
A pass on the tests and the implementations of the methods.

I still need to review `sklearn_numba_dpex/kmeans/kernels/compute_{euclidean_distances,inertia,labels}.py`.

File: sklearn_numba_dpex/kmeans/kernels.py (hunk 1 above)
Comment (jjerphan) [reply sklearn_numba_dpex/kmeans/kernels.py]: > While I agree with your rule of thumbs for diffs, i don't think it should apply here since the diff was artificially high because of copy/paste, even if they appear in the diff I don't think the internals of the copy/pasted blocks should be reviewed, they have not been edited, only moved left and right.

I agree, but to me, having blocks of code moved around without changes still hardened reviews. PR like #18 highly help in this regards.

File: sklearn_numba_dpex/kmeans/drivers.py (hunk 9 above)
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/drivers.py:563]: We monkey patch sklearn internals with this function in places where it's always called with `squared=False`.

This check ensures that this assumption holds.

Could as well be `RuntimeError` or `ValueError`.

I didn't try to make the message more explicit on purpose: all this monkey-patching salad is going to be removed in a soon to come PR in favor of the more conventional plugin interface.

File: sklearn_numba_dpex/kmeans/drivers.py (hunk 8 above)
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/drivers.py:551]: no actually I'm wrong: the current code is good, `reduce_inertia_kernel(per_sample_inertia)` is a 1-sized array

File: sklearn_numba_dpex/kmeans/drivers.py (hunk 9 above)
Comment (jjerphan) [reply sklearn_numba_dpex/kmeans/drivers.py:567]: OK, then if supporting `squared=True` is possible but useless, I think it is worth adding an inline comment to mention it. For instance:
```suggestion
        if squared:
            # This implementation is used to monkey patch sklearn private
            # functions' which are always called with `squared=False`.
            # It's possible to support `squared=True`, but useless in this
            # context, hence we raise an error.
            raise NotImplementedError("Only squared=False is allowed")
```

File: sklearn_numba_dpex/kmeans/drivers.py (hunk 8 above)
Comment (jjerphan) [reply sklearn_numba_dpex/kmeans/drivers.py]: In this case, I think having a comment indicates that it is a 1-sized array whose scalar is accessed via `[0]` would be useful for readers.

File: sklearn_numba_dpex/kmeans/kernels/compute_labels.py [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+                dot_products,
+            )
+
+            dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+
+            first_centroid_idx += window_n_centroids
+
+        # No update step, only store min_idx in the output array
+        if sample_idx >= n_samples:
+            return
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/kernels/compute_labels.py:142]: Shouldn't this be moved at before for-loops?

File: sklearn_numba_dpex/kmeans/kernels/compute_inertia.py [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+                centroids_window_half_l2_norm,
+                dot_products,
+            )
+
+            dpex.barrier(dpex.CLK_LOCAL_MEM_FENCE)
+
+            first_centroid_idx += window_n_centroids
+
+        if sample_idx >= n_samples:
+            return
```
Comment (jjerphan) [inline sklearn_numba_dpex/kmeans/kernels/compute_inertia.py:160]: Shouldn't this be moved at before for-loops?

**Review by jjerphan — APPROVED** [review]
LGTM. Just a remark.

File: sklearn_numba_dpex/kmeans/kernels/compute_labels.py (hunk 14 above)
Comment (fcharras) [reply sklearn_numba_dpex/kmeans/kernels/compute_labels.py:142]: ah this is a good question and actually a subtle point that would benefit from a comment, IIRC I found that an undesirable work item (such as `sample_idx >= n_samples`) still has to reach the barriers, else the kernel will deadlock.

In general kernels with such code

```python
If some_condition:
    do_sth()
    dpex.barrier(...)
else:
    maybe_do_sth_else()
```

will deadlock, and must be written like this:

```python
If some_condition:
    do_sth()
else:
    maybe_do_sth_else().
    # no barrier

dpex.barrier(...)
```

Comment (jjerphan) [reply sklearn_numba_dpex/kmeans/kernels/compute_labels.py:142]: OK, I think this better be made explicit in a comment as you suggest.

**Comment by fcharras** [comment]
Rebased to show co authorship in review commits I think it worked ?

**Comment by jjerphan** [comment]
Yes, it worked. Commits are marked as unverified when their (co-)author usually GPG-sign them.
